lib/active_merchant/billing/gateways/nuvei.rb

```diff
@@ -0,0 +1,137 @@
+  module Billing
+    class NuveiGateway < Gateway
+      self.test_url = 'https://ppp-test.nuvei.com/ppp/api/v1'
+      self.live_url = 'https://secure.safecharge.com/ppp/api/v1'
+
+      self.supported_countries = %w[US CA IN NZ GB AU US]
+      self.default_currency = 'USD'
+      self.money_format = :cents
+      self.supported_cardtypes = %i[visa master american_express discover union_pay]
+      self.currencies_without_fractions = %w[CLP KRW JPY ISK MMK PYG UGX VND XAF XOF]
+      self.homepage_url = 'https://www.nuvei.com/'
+      self.display_name = 'Nuvei'
+
+      ENDPOINTS_MAPPING = {
+        authenticate: '/getSessionToken',
+        purchase: '/payment', # /authorize with transactionType: "Auth"
+        capture: '/settleTransaction',
+        refund: '/refundTransaction',
+        void: '/voidTransaction',
+        general_credit: '/payout'
+      }
+
+      def initialize(options = {})
+        requires!(options, :merchant_id, :merchant_site_id, :secret_key)
+        super
+      end
+
+      def authorize(money, payment, options = {}); end
+
+      def purchase(money, payment, options = {}); end
+
+      def capture(money, authorization, options = {}); end
+
+      def refund(money, authorization, options = {}); end
+
+      def void(authorization, options = {}); end
+
+      def credit(money, payment, options = {}); end
+
+      def scrub(transcript)
+        transcript.
+          gsub(%r((Authorization: Bearer )[a-zA-Z0-9._-]+)i, '\1[FILTERED]').
+          gsub(%r((\"cardNumber\\\":\\\")\d+), '\1[FILTERED]').
+          gsub(%r((\"cardCvv\\\":\\\")\d+), '\1[FILTERED]').
+          gsub(%r((\"cardHolderName\\\":\\\")\w+), '\1[FILTERED]')
+      end
+
+      private
+
+      def add_payment(post, payment); end
+
+      def commit(action, post, authorization = nil, method = :post)
+        MultiResponse.run do |r|
+          r.process { fetch_session_token } unless session_token_valid?
+          r.process do
+            api_request(action, post, authorization, method).tap do |response|
+              response.params.merge!(@options.slice(:session_token, :token_expires)) if @options[:new_credentials]
+            end
+          end
+        end
+      end
+
+      def session_token_valid?
+        @options[:session_token] && @options[:token_expires] && Time.now <= @options[:token_expires]
+      end
+
+      def fetch_session_token
+        post = {
+          'merchantId' => @options[:merchant_id],
+          'merchantSiteId' => @options[:merchant_site_id],
+          'merchantKey' => @options[:secret_key],
+          'clientRequestId' => SecureRandom.uuid,
+          'timeStamp' => Time.now.utc.strftime('%Y%m%d%H%M%S')
+        }
+
+        post.tap { |p| p['checksum'] = calculate_checksum(p) }
+
+        response = JSON.parse(ssl_post(url(:authenticate), post.to_json, headers)).with_indifferent_access
+        current_time = Time.now
+        expiration_time = current_time + (15 * 60) # 15 minutes in seconds
```

Comment:
⚠️ Warning:

What I can see on the intention is that current_time should be a Unix Epoch number but here you are storing a date, so in the future you are going to compare the string "2024-07-18 15:31:03.326644 -0500" witht epoch time that `DateTime.now.strftime('%Q').to_i` produces

![tmux](https://github.com/user-attachments/assets/d8085983-20dc-49b7-8d90-eb9ed0d11b9f)

Reply:
good Catch , thanks I will update it